jesd204,adrv9009: add support for stop states and resuming from them

commodo opened the pull request:
This adds support for stop-states in the FSM.

For the 9009 we need to hack it a bit to re-use libiio to query when the FSM has paused into a
certain state and to resume from that state.

With this, in a device tree, any JESD204 device can define some states on
which it want to defer.

Example:
       jesd204-stop-states = <JESD204_FSM_STATE_LINK_PRE_SETUP
                              JESD204_FSM_STATE_LINK_ENABLE>;

The information can then be queried:

iio_attr -q -a -d adrv9009-phy jesd204_fsm_state
iio_attr -q -a -d adrv9009-phy jesd204_fsm_error
iio_attr -q -a -d adrv9009-phy jesd204_fsm_paused

And then a resume can be called:

iio_attr -q -a -d adrv9009-phy jesd204_fsm_resume 1

Signed-off-by: Alexandru Ardelean <alexandru.ardelean@analog.com>

nunojsa commented on drivers/jesd204/jesd204-priv.h:
```diff
@@ -125,6 +128,7 @@ struct jesd204_dev {
 	bool				is_top;
 
 	bool				is_sysref_provider;
+	bool				stop_states[JESD204_FSM_STATES_NUM + 1];
```
won't checkpatch complain about `bool`?

commodo replied:
maybe in newer versions of checkpatch;
but i'll probably address them later; maybe with this new BITMAP() type

nunojsa commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -682,11 +714,19 @@ static int jesd204_validate_fsm_data(struct jesd204_dev *jdev,
 
 	if (link_idx != JESD204_LINKS_ALL) {
 		ol = &jdev_top->active_links[link_idx];
```
nit: extra line?

nunojsa commented on drivers/jesd204/jesd204-fsm.c:
```diff
@@ -682,11 +714,19 @@ static int jesd204_validate_fsm_data(struct jesd204_dev *jdev,
 		ol = &jdev_top->active_links[link_idx];
+		ret = jesd204_validate_resuming_state(jdev, ol, fsm_data);
+		if (ret)
+			return ret;
```
nit: extra line?

commodo replied:
ack

nunojsa commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -33,6 +33,52 @@ static unsigned int jesd204_con_id_counter;
 
+int jesd204_get_links_data(struct jesd204_dev *jdev,
+			   struct jesd204_link ** const links,
+			   unsigned int num_links)
```
`num_links` should be `const`...

commodo replied:
here, it doesn't make much difference;
it's a copy on the stack; and sometimes, i like to use these ints, instead of defining a new variable inside;
so, i can add a `const`, but not much added benefit [for now]

nunojsa replied:
I think it's always good from an API point of view to define things as they are. Also, `const` let's the compiler do some optimization that we might not realize... Either way, up to you :smile:

nunojsa commented on drivers/jesd204/jesd204-core.c:
```diff
@@ -33,6 +33,52 @@ static unsigned int jesd204_con_id_counter;
+
+bool jesd204_link_get_paused(const struct jesd204_link *lnk)
+{
+	struct jesd204_link_opaque *ol = container_of(lnk, typeof(*ol), link);
```
nit: Personal preference, explicitly using `struct jesd204_link_opaque` is more readable to me but also fine to use `typeof`

commodo replied:
i also thought about it;
will change;
thing is, it makes the line long;

nunojsa replied:
It actually came to my mind that could be reason for the `typeof`. Either way, we should already just thing in 100 col lines (don't know if we pass that without `typeof`)

nunojsa reviewed (approved):
LGTM. Just some minor observations...

commodo commented:
Changelog v1 -> v2:
* added `const` to num_links in jesd204_get_links_data()
* added extra lines jesd204_validate_fsm_data()
* ` container_of(lnk, typeof(*ol), link);` -> `container_of(lnk, struct jesd204_link_opaque, link);`
